Deflake PartitionReassignmentsTest

joe-redpanda opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* None

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/redpanda-data/redpanda?pullRequest=26452) <br/>All committers have signed the CLA.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/redpanda-data/redpanda?pullRequest=26452) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/redpanda-data/redpanda?pullRequest=26452) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/redpanda-data/redpanda?pullRequest=26452) it.</sub>

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/67297-->
<!--b8fda178-5182-4bbf-8c17-1e4a5267199d-->
<!--title-->
## Retry command for [Build#67297](https://buildkite.com/redpanda/redpanda/builds/67297)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/partition_reassignments_test.py::PartitionReassignmentsTest.test_reassignments
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--abf6f14b-49b8-41b1-9b8c-53e74fe5184d-->
<!--TEST RESULTS-->
<!--3822ddee-5aa0-4802-b87d-620131eab7a1-->
<!--TEST RESULTS-->
<!--8924fff1-7add-41d4-9e8c-ebeb605af742-->
<!--TEST RESULTS-->
<!--01a80d3a-1394-494d-b738-44830db507e7-->
<!--TEST RESULTS-->
<!--ac58dfd6-a332-4e50-a986-f13cfb05aa72-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67297">build#67297</a></summary>


| test_class                  | test_method                | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------------|:---------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| PartitionBalancerTest       | test_fuzz_admin_ops        |                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67297#019766ea-b463-40bd-99e1-ecb473ad10e5 | FLAKY         | 20/21    | upstream reliability is '96.05263157894737'. current run reliability is '95.23809523809523'. drift is 0.81454 and the allowed drift is set to 50. The test should PASS |
| PartitionReassignmentsTest  | test_reassignments         |                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67297#019766f8-513f-4fe8-bd0f-698724f8feba | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test | {"cloud_storage_type": 2} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67297#019766f8-513f-4fe8-bd0f-698724f8feba | FLAKY         | 17/21    | upstream reliability is '100.0'. current run reliability is '80.95238095238095'. drift is 19.04762 and the allowed drift is set to 50. The test should PASS            |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67351">build#67351</a></summary>


| test_class                                   | test_method                                  | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:---------------------------------------------|:---------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| MaintenanceTest                              | test_maintenance_sticky                      | {"use_rpk": false}                                                                                                                             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67351#01976ba2-2a18-4079-b234-f8e66b2b1b83 | FLAKY         | 19/21    | upstream reliability is '96.46017699115043'. current run reliability is '90.47619047619048'. drift is 5.98399 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest                     | test_node_operations                         | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67351#01976ba2-2a19-4f56-a7af-cdd478250df4 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| src/v/crypto/tests/crypto_bench_rpbench_test | src/v/crypto/tests/crypto_bench_rpbench_test |                                                                                                                                                | unit        | https://buildkite.com/redpanda/redpanda/builds/67351#01976b6f-5f13-4e0c-abed-c73022a131dc | FAIL          | 0/1      |                                                                                                                                                                        |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67411">build#67411</a></summary>


| test_class                  | test_method                              | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:----------------------------|:-----------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ConsumerOffsetsRecoveryTest | test_consumer_offsets_partition_recovery |                                                                                                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67411#01977ad6-e278-45f0-ae3f-6fd48df40c32 | FLAKY         | 19/21    | upstream reliability is '97.5'. current run reliability is '90.47619047619048'. drift is 7.02381 and the allowed drift is set to 50. The test should PASS               |
| RaftAvailabilityTest        | test_controller_node_isolation           |                                                                                                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67411#01977ad6-e279-43c0-82c9-3176367cc5ab | FLAKY         | 20/21    | upstream reliability is '94.82758620689656'. current run reliability is '95.23809523809523'. drift is -0.41051 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest    | test_node_operations                     | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67411#01977ad6-e278-45f0-ae3f-6fd48df40c32 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
| DisablingPartitionsTest     | test_disable                             |                                                                                                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67411#01977af3-7e92-434b-9f59-cdeae76cd812 | FLAKY         | 16/21    | upstream reliability is '94.00428265524626'. current run reliability is '76.19047619047619'. drift is 17.81381 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67577">build#67577</a></summary>


| test_class                  | test_method                | test_arguments                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:----------------------------|:---------------------------|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| IcebergUsageTest            | test_iceberg_usage         | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "spark"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67577#0197847b-1fa1-4045-b033-4c4c19cf1e58 | FLAKY         | 16/21    | upstream reliability is '84.5'. current run reliability is '76.19047619047619'. drift is 8.30952 and the allowed drift is set to 50. The test should PASS             |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test | {"cloud_storage_type": 1}                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67577#0197847b-1fa1-4045-b033-4c4c19cf1e58 | FLAKY         | 20/21    | upstream reliability is '98.09069212410502'. current run reliability is '95.23809523809523'. drift is 2.8526 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67665">build#67665</a></summary>


| test_class                 | test_method                              | test_arguments                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:---------------------------|:-----------------------------------------|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| IcebergUsageTest           | test_iceberg_usage                       | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "spark"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67665#01978e57-2e83-4401-b216-122d037bc37e | FLAKY         | 19/21    | upstream reliability is '85.3035143769968'. current run reliability is '90.47619047619048'. drift is -5.17268 and the allowed drift is set to 50. The test should PASS |
| TxAtomicProduceConsumeTest | test_basic_tx_consumer_transform_produce | {"with_failures": true}                                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67665#01978e57-2e83-4401-b216-122d037bc37e | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
</details>
<!--/body-->

bashtanov commented:
After [this change merged 10 days ago](https://github.com/redpanda-data/redpanda/pull/26149) the test will retry the partition reassignment if clashing with partition balancer. From what I can see [in one of the test logs](https://ci-artifacts.dev.vectorized.cloud/redpanda/67206/01976065-3f40-47a6-8440-e9546c454aa8/vbuild/ducktape/results/2025-06-11--001/PartitionReassignmentsTest/test_add_partitions_with_inprogress_reassignments/118/test_log.debug) there are concurrent reassignment(s) in progress for 10 seconds. That sounds a bit too long to me, but to be on the safe side I'd maybe give it 30 seconds as partition balancer may need to move it multiple times to achieve a stable state. @ztlpn might give a better advice on how long we expect it to take. If it still reproduces I'd investigate whether it's actually partition balancer (did you look in broker logs?), and why it takes so long. If partition balancer won't stop moving it then it's a bug, although not the kind of a bug this test is checking.

> This test only requires that all partitions be currently reassigning to perform its function.

That's true. However, the partition balancer action may be almost done when we attempt a manual reassignment, and when we add partitions it's complete. This will make the test fail. Have you tried running it 100-1000 times with your change to see if it is stable?

bashtanov commented:
Please split into appropriately annotated commits as per https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#commit-history

joe-redpanda commented:
> Please split into appropriately annotated commits as per https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#commit-history

squash merged, should be fixed

mmaslankaprv commented on tests/rptest/tests/partition_reassignments_test.py:
```diff
@@ -188,7 +188,11 @@ def __init__(self, test_context):
+                             extra_rp_conf={
+                                 "enable_leader_balancer": False,
+                                 "partition_autobalancing_mode": "off",
+                             })
```
Can you elaborate a bit in the commit message why balancer is influencing the test outcome ?

mmaslankaprv commented:
please prefix the commit message with the area it is related with, in this case it will be `tests: ...`

joe-redpanda commented:
/ci-repeat 1
tests/rptest/tests/partition_reassignments_test.py::PartitionReassignmentsTest.test_reassignments

bashtanov commented on tests/rptest/tests/partition_reassignments_test.py:
```diff
@@ -547,7 +554,7 @@ def add_partition_during_inprogress_reassignment(
             assert try_add_partitions(
                 topic, count
-            ) is None, f'Expected failed add-partitions: topic {topic}, partition count {count}, output {out}'
+            ) is None or str, f'Expected reassignment in progress or success: topic {topic}, partition count {count}'
```
This assert seems to always pass. It is parsed as `(try_add_partitions(topic, count) is None) or str`. `str` is a class, and a python class is an object of class `type`. The latter does not implement `__bool__(self)` or `__len__(self)` methods, so its instances are deemed true when used in a boolean context.

Perhaps you meant to check if the returned object is either None or an instance of `str`? `is` is not about checking the type, it's checking identity, not unlike `===` in other languages. So it'd be `res is None or isinstance(res, str)`.

joe-redpanda replied:
thats a good catch. Let me remap this. Assert is probably not the right call here.

dotnwat reviewed (commented):
> tests: deflake partition_reassignments_test.py by disabling autobalance

Please describe in the commit message why autobalancing was causing the problem and why it's the right solution for deflaking the test.

Also, in the cover letter I don't think it's really necessary to add any release notes.

Commit:
tests: deflake partition_reassignments_test.py

test_add_partitions_with_inprogress_reassignments attempts to add a
partition to a topic for which all partitions are undergoing
reassignment.

It accomplished this by issuing a reassignment to all paritions
manually, and checking the success of these commands. These commands
could fail if the autobalancer starts a reassignment for the sake of
balancing the cluster, because a reassignment would already be in
progress.

This change disables the autobalancer for this test such that the manual
reassignment command will not race the autobalancer.

This also allows the test to succeed if adding a partition succeeds, as
the failure behavior was written to keep parity with Kafka. This may be
changed in the future to simply succeed.

bashtanov commented on tests/rptest/tests/partition_reassignments_test.py:
```diff
@@ -531,23 +537,28 @@ def test_add_partitions_with_inprogress_reassignments(self):
         # Expect fail.
         def add_partition_during_inprogress_reassignment(
                 topic: str, count: int):
-            assert try_add_partitions(
-                topic, count
-            ) is None, f'Expected failed add-partitions: topic {topic}, partition count {count}, output {out}'
```
The idea of this test is to make sure the operation fails with this particular error.

This assert actually checked that RPK did throw with `REASSIGNMENT_IN_PROGRESS`. If`rpk.add_partitions` succeeds it returns what `rpk` printed to stdout, and even if it didn't print a thing it would be an empty string, not None -- that's how Python's Popen works.

joe-redpanda replied:
Per discussion in scrum, we considered removing this test because the behavior to deny a new partition while reassignments were in progress was solely to keep parity with kafka.

RP has no issue adding a partition with an ongoing reassignment

Rather than delete the test though, we decided that we could just future proof the test to accept either REASSIGNMENT_IN_PROGRESS error or success

success -> str
(with updates) REASSIGNMENT -> str
all others -> exception, which will fail the test

am I missing something?
